### PR TITLE
Adds temperature and pressure sensors to circuits

### PR DESCRIPTION
**Pull request by read-0nly**
## About The Pull Request

Adds two components to circuits - temperature and pressure sensor.

## Why It's Good For The Game

Allows circuits to react to environmental triggers - for instance, tie it to a speech module, leave a radio nearby, toss both near the engine, engine station-wide fire alarm.

## Changelog
:cl:
add: Added temperature and pressure sensors to circuits
/:cl:

### CODE REVIEW

**Comment by read-0nly** [comment]
Other ideas I'm hoping to implement:
- Light sensor
- Radiation sensor
- Specific gas sensors (each their own component was the idea, open to a unified sensor)
- Extinguisher unit (thought you might need to feed it an extinguisher, crowbar to remove, sprays from one tile forward, facing backwards - for auto-extinguishers)

**Comment by optimumtact** [comment]
To ensure you don't run afoul of maintainers, please make sure you have read and understood the integrated circuits design document. https://hackmd.io/@tgstation/HyOpC1ZjP

Most of what you proposed should be fine.

**Comment by read-0nly** [comment]
@optimumtact thanks for the heads up! Bookmarked, I'll make sure to look at it as I'm working through these ideas!

File: code/modules/research/techweb/all_nodes.dm [hunk 1]
```diff
@@ -224,6 +224,8 @@
 		"comp_speech",
 		"comp_speech",
 		"comp_string_contains",
+		"comp_tempsensor",
+		"comp_pressuresensor",
```
Comment (Watermelon914) [inline code/modules/research/techweb/all_nodes.dm:228]: these should be in alphabetical order, so comp_tempsensor should be around components starting with "comp_t", similar thing for comp_pressuresensor.

File: code/modules/wiremod/components/sensors/pressuresensor.dm [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	if(.)
+		return
+
+	if(!COMPONENT_TRIGGERED_BY(trigger, port))
+		return
+	var/turf/location = parent.shell.loc
+	if(!location)
+		location=parent.loc
+	if(!location)
+		location=loc
```
Comment (Watermelon914) [inline code/modules/wiremod/components/sensors/pressuresensor.dm:37]:
```suggestion
	var/turf/location = get_turf(src)
	if(!location)
		result.set_output(null)
		return
```

File: code/modules/wiremod/components/sensors/pressuresensor.dm [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	var/datum/gas_mixture/environment = location.return_air()
+	var/total_moles = environment.total_moles()
+	var/pressure = environment.return_pressure()
+	if(total_moles)
+		//environment.assert_gases(arglist(GLOB.hardcoded_gases))
+		result.set_output(round(pressure,1))
```
Comment (Watermelon914) [inline code/modules/wiremod/components/sensors/pressuresensor.dm:43]: You'll want to set an output if there isn't any moles to indicate that it has no pressure/the result wasn't valid.

File: code/modules/wiremod/components/sensors/tempsensor.dm [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * # Temperature Sensor
+ *
+ * Returns the temperature of the tile
+ */
+/obj/item/circuit_component/tempsensor
+	display_name = "Temperature Sensor"
+	display_desc = "Outputs the current temperature of the tile"
+
+	/// Trigger a measurement
+	var/datum/port/input/trigger
```
Comment (Watermelon914) [inline code/modules/wiremod/components/sensors/tempsensor.dm:11]:
```suggestion
	circuit_flags = CIRCUIT_FLAG_INPUT_SIGNAL|CIRCUIT_FLAG_OUTPUT_SIGNAL
```

File: code/modules/wiremod/components/sensors/pressuresensor.dm [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ * # Pressure Sensor
+ *
+ * Returns the pressure of the tile
+ */
+/obj/item/circuit_component/pressuresensor
+	display_name = "Pressure Sensor"
+	display_desc = "Outputs the current pressure of the tile"
+
+	/// Trigger a measurement
+	var/datum/port/input/trigger
```
Comment (Watermelon914) [inline code/modules/wiremod/components/sensors/pressuresensor.dm:11]: Don't need this, there are flags that handle this for you.
```suggestion
	circuit_flags = CIRCUIT_FLAG_INPUT_SIGNAL|CIRCUIT_FLAG_OUTPUT_SIGNAL
```

File: code/modules/wiremod/components/sensors/tempsensor.dm [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	if(.)
+		return
+
+	if(!COMPONENT_TRIGGERED_BY(trigger, port))
+		return
+	var/turf/location = parent.shell.loc
+	if(!location)
+		location=parent.loc
+	if(!location)
+		location=loc
```
Comment (Watermelon914) [inline code/modules/wiremod/components/sensors/tempsensor.dm:37]:
```suggestion
	var/turf/location = get_turf(src)
	if(!location)
		result.set_output(null)
		return
```

File: code/modules/wiremod/components/sensors/tempsensor.dm [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	var/datum/gas_mixture/environment = location.return_air()
+	var/total_moles = environment.total_moles()
+	if(total_moles)
+		//environment.assert_gases(arglist(GLOB.hardcoded_gases))
+		result.set_output(round(environment.temperature,1))
```
Comment (Watermelon914) [inline code/modules/wiremod/components/sensors/tempsensor.dm:42]: Same as what I said above here.

File: code/modules/research/techweb/all_nodes.dm (hunk 1 above)
Comment (Watermelon914) [reply code/modules/research/techweb/all_nodes.dm:228]: comp_tempsensor is in a good spot actually but comp_pressuresensor needs to be moved

**Comment by Watermelon914** [comment]
> * Specific gas sensors (each their own component was the idea, open to a unified sensor)

Options do exist so that you can easily have this as 1 component instead of having many different components for it.

> * Interface to tie to igniters, timers, health sensors, etc as triggers

Health sensors do already exists and you can trigger igniters via signallers if they're in a signaller/igniter assembly

> * Some sort of gas venter thing - you feed it a tank, then can trigger the release of the gas. Maybe you stuff it with cold gasses and sprinkle them as auto fire-stoppers or something, I dunno, I didn't think this one through much.

If you've seen usb cables yet, you can basically use these to connect to physical structures, one of which being a digital vent, so you can already do this.

**Comment by Watermelon914** [comment]
Just letting you know that you're also going to have to manually go into the code and remove your references to `trigger`. Also do remember to test to make sure it works because I'm not really able to test your code.

**Comment by godspeedfunnyman** [comment]
looks cool, you can make smart firelocks with this (closes a door and bolts it and you can make it unbolt it using voice or just closes a door)
also you should make us able to connect USB to an air alarm and get information/control it, like the security console one

**Comment by Watermelon914** [comment]
LGTM

**Comment by Mothblocks** [comment]
> Health sensors do already exists and you can trigger igniters via signallers if they're in a signaller/igniter assembly

TBF, I think circuits have the ability to replace the more rudimentary pre-components @Watermelon914

**Pull request merged**